File: sdks/typescript/adapter.mjs
#!/usr/bin/env node
/**
 * Conformance adapter for the TypeScript/JS SDK — rcx-verify-contract/1 §7.
 *
 * Newline-delimited JSON on stdin, one response per line on stdout, in order.
 * Malformed input is a verdict, never a crash and never a non-zero exit.
 *
 *   ./scripts/conformance-harness.py --adapter "node sdks/typescript/adapter.mjs"
 */

import { createInterface } from 'node:readline';

import {
  CHAIN_ENTRY_ENRICHED,
  CHAIN_SNAPSHOT,
  SnapshotEntry,
  VerifyError,
  verifyHistory,
  verifyNamespace,
  verifyPublisher,
  verifyReceipt,
  verifySnapshot,
} from './index.mjs';

const invalid = (code, detail = null) => ({ valid: false, error: code, detail });

const fromError = (error) =>
  error instanceof VerifyError
    ? { valid: false, error: error.code, detail: error.detail }
    : invalid('decode_error', `${error.name}: ${error.message}`);

function unhex(request, key, expectedLength = null) {
  const raw = request?.[key];
  if (typeof raw !== 'string' || raw.length % 2 !== 0 || /[^0-9a-fA-F]/.test(raw)) return null;
  const bytes = new Uint8Array(raw.length / 2);
  for (let index = 0; index < bytes.length; index += 1) {
    bytes[index] = Number.parseInt(raw.slice(index * 2, index * 2 + 2), 16);
  }
  if (expectedLength !== null && bytes.length !== expectedLength) return null;
  return bytes;
}

const toHex = (bytes) =>
  Array.from(bytes, (byte) => byte.toString(16).padStart(2, '0')).join('');

const HANDLERS = {
  verifyReceipt(request) {
    const cbor = unhex(request, 'signedCanonicalCborHex');
    const key = unhex(request, 'publicKeyHex');
    if (!cbor || !key) return invalid('decode_error', 'bad hex input');
    const facts = verifyReceipt(cbor, key);
    return {
      valid: true,
      receiptHashHex: toHex(facts.receiptHash),
      signerKidPresent: Boolean(facts.signerKid),
      snapshotRootHex: facts.snapshotRoot ? toHex(facts.snapshotRoot) : null,
    };
  },

  verifySnapshot(request) {
    const expected = unhex(request, 'expectedRootHex', 32);
    if (!expected) return invalid('decode_error', 'bad expectedRootHex');
    if (!Array.isArray(request.entries)) return invalid('decode_error', 'missing entries');
    const entries = [];
    for (const item of request.entries) {
      const { name, version, canonicalJson } = item ?? {};
      if (
        typeof name !== 'string' ||
        typeof version !== 'string' ||
        typeof canonicalJson !== 'string'
      ) {
        return invalid('decode_error', 'malformed entry');
      }
      entries.push(new SnapshotEntry(name, version, canonicalJson));
    }
    verifySnapshot(entries, expected);
    return { valid: true };
  },

  verifyPublisher(request) {
    const expected = unhex(request, 'expectedDeclaredHashHex', 32);
    const declaration = request.declarationJson;
    if (!expected || typeof declaration !== 'string') {
      return invalid('decode_error', 'missing declarationJson or hash');
    }
    verifyPublisher(declaration, expected);
    return { valid: true };
  },

  verifyNamespace(request) {
    const expected = unhex(request, 'expectedDeclaredHashHex', 32);
    const declaration = request.declarationJson;
    const namespace = request.claimedNamespace;
    if (!expected || typeof declaration !== 'string' || typeof namespace !== 'string') {
      return invalid('decode_error', 'missing namespace inputs');
    }
    verifyNamespace(declaration, expected, namespace);
    return { valid: true };
  },

  verifyHistory(request) {
    const key = unhex(request, 'publicKeyHex');
    if (!key) return invalid('decode_error', 'bad publicKeyHex');
    if (request.kind !== CHAIN_SNAPSHOT && request.kind !== CHAIN_ENTRY_ENRICHED) {
      return invalid('decode_error', 'kind must be snapshot|entryEnriched');
    }
    if (!Array.isArray(request.linksHex)) return invalid('decode_error', 'missing linksHex');
    const links = [];
    for (const item of request.linksHex) {
      const bytes = unhex({ item }, 'item');
      if (!bytes) return invalid('decode_error', 'malformed link hex');
      links.push(bytes);
    }
    verifyHistory(links, key, request.kind);
    return { valid: true };
  },
};

const readline = createInterface({ input: process.stdin, crlfDelay: Infinity });

for await (const line of readline) {
  if (!line.trim()) continue;

  let response;
  let request;
  try {
    request = JSON.parse(line);
  } catch (error) {
    process.stdout.write(`${JSON.stringify(invalid('decode_error', error.message))}\n`);
    continue;
  }

  const handler = typeof request?.verb === 'string' ? HANDLERS[request.verb] : undefined;
  if (!handler) {
    const detail = request?.verb ? `unknown verb: ${request.verb}` : 'missing verb';
    process.stdout.write(`${JSON.stringify(invalid('decode_error', detail))}\n`);
    continue;
  }

  try {
    response = handler(request);
  } catch (error) {
    // A thrown error must still be a verdict — the harness has to distinguish
    // "this SDK says invalid" from "this SDK fell over".
    response = fromError(error);
  }
  process.stdout.write(`${JSON.stringify(response)}\n`);
}

File: sdks/python/adapter.py
#!/usr/bin/env python3
"""Conformance adapter for the Python SDK — rcx-verify-contract/1 §7.

Newline-delimited JSON on stdin, one response per line on stdout, in order.
Malformed input is a verdict, never a crash and never a non-zero exit.

    ./scripts/conformance-harness.py --adapter "python3 sdks/python/adapter.py"
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from rcx_verify import (  # noqa: E402
    CHAIN_ENTRY_ENRICHED,
    CHAIN_SNAPSHOT,
    SnapshotEntry,
    VerifyError,
    verify_history,
    verify_namespace,
    verify_publisher,
    verify_receipt,
    verify_snapshot,
)


def invalid(code: str, detail: str = "") -> dict:
    return {"valid": False, "error": code, "detail": detail}


def unhex(request: dict, key: str, expected_len: int | None = None):
    raw = request.get(key)
    if not isinstance(raw, str):
        return None
    try:
        value = bytes.fromhex(raw)
    except ValueError:
        return None
    if expected_len is not None and len(value) != expected_len:
        return None
    return value


def do_verify_receipt(request: dict) -> dict:
    cbor = unhex(request, "signedCanonicalCborHex")
    key = unhex(request, "publicKeyHex")
    if cbor is None or key is None:
        return invalid("decode_error", "bad hex input")
    try:
        facts = verify_receipt(cbor, key)
    except VerifyError as exc:
        return {"valid": False, "error": exc.code, "detail": exc.detail}
    return {
        "valid": True,
        "receiptHashHex": facts.receipt_hash.hex(),
        "signerKidPresent": bool(facts.signer_kid),
        "snapshotRootHex": facts.snapshot_root.hex() if facts.snapshot_root else None,
    }


def do_verify_snapshot(request: dict) -> dict:
    expected = unhex(request, "expectedRootHex", 32)
    if expected is None:
        return invalid("decode_error", "bad expectedRootHex")
    raw_entries = request.get("entries")
    if not isinstance(raw_entries, list):
        return invalid("decode_error", "missing entries")
    entries = []
    for item in raw_entries:
        if not isinstance(item, dict):
            return invalid("decode_error", "malformed entry")
        name, version, canonical = (
            item.get("name"),
            item.get("version"),
            item.get("canonicalJson"),
        )
        if not all(isinstance(f, str) for f in (name, version, canonical)):
            return invalid("decode_error", "malformed entry")
        entries.append(SnapshotEntry(name, version, canonical))
    try:
        verify_snapshot(entries, expected)
    except VerifyError as exc:
        return {"valid": False, "error": exc.code, "detail": exc.detail}
    return {"valid": True}


def do_verify_publisher(request: dict) -> dict:
    expected = unhex(request, "expectedDeclaredHashHex", 32)
    if expected is None or "declaration" not in request:
        return invalid("decode_error", "missing declaration or hash")
    try:
        verify_publisher(request["declaration"], expected)
    except VerifyError as exc:
        return {"valid": False, "error": exc.code, "detail": exc.detail}
    return {"valid": True}


def do_verify_namespace(request: dict) -> dict:
    expected = unhex(request, "expectedDeclaredHashHex", 32)
    namespace = request.get("claimedNamespace")
    if expected is None or "declaration" not in request or not isinstance(namespace, str):
        return invalid("decode_error", "missing namespace inputs")
    try:
        verify_namespace(request["declaration"], expected, namespace)
    except VerifyError as exc:
        return {"valid": False, "error": exc.code, "detail": exc.detail}
    return {"valid": True}


def do_verify_history(request: dict) -> dict:
    key = unhex(request, "publicKeyHex")
    if key is None:
        return invalid("decode_error", "bad publicKeyHex")
    kind = request.get("kind")
    if kind not in (CHAIN_SNAPSHOT, CHAIN_ENTRY_ENRICHED):
        return invalid("decode_error", "kind must be snapshot|entryEnriched")
    raw_links = request.get("linksHex")
    if not isinstance(raw_links, list):
        return invalid("decode_error", "missing linksHex")
    links = []
    for item in raw_links:
        if not isinstance(item, str):
            return invalid("decode_error", "malformed link hex")
        try:
            links.append(bytes.fromhex(item))
        except ValueError:
            return invalid("decode_error", "malformed link hex")
    try:
        verify_history(links, key, kind)
    except VerifyError as exc:
        return {"valid": False, "error": exc.code, "detail": exc.detail}
    return {"valid": True}


DISPATCH = {
    "verifyReceipt": do_verify_receipt,
    "verifySnapshot": do_verify_snapshot,
    "verifyPublisher": do_verify_publisher,
    "verifyNamespace": do_verify_namespace,
    "verifyHistory": do_verify_history,
}


def main() -> int:
    for line in sys.stdin:
        if not line.strip():
            continue
        try:
            request = json.loads(line)
        except json.JSONDecodeError as exc:
            print(json.dumps(invalid("decode_error", str(exc))), flush=True)
            continue

        verb = request.get("verb") if isinstance(request, dict) else None
        handler = DISPATCH.get(verb)
        if handler is None:
            detail = f"unknown verb: {verb}" if verb else "missing verb"
            print(json.dumps(invalid("decode_error", detail)), flush=True)
            continue

        try:
            response = handler(request)
        except Exception as exc:  # noqa: BLE001 — a crash must still be a verdict
            response = invalid("decode_error", f"{type(exc).__name__}: {exc}")
        print(json.dumps(response), flush=True)
    return 0


if __name__ == "__main__":
    sys.exit(main())
